qcodes/instrument/base.py

```diff
@@ -401,6 +424,259 @@ def validate_status(self, verbose: bool = False) -> None:
+            # the `@add_parameter` decorator adds an attribute we check for here
+            if hasattr(meth, _ADD_PARAMETER_ATTR_NAME):
+                param_name = meth.__name__[len(_DECORATED_METHOD_PREFIX):]
+                kwargs = {
+                    kw_name: kwarg_to_attr(self, kw_name, kw_value.default, param_name)
+                    for kw_name, kw_value in inspect.signature(meth).parameters.items()
+                }
+                if "docstring" in kwargs:
+                    raise RuntimeError(
+                        f"Failed to add parameter {param_name!r}.\n"
+                        f"`docstring` kwarg was provided to {meth.__name__!r} method. "
+                        f"Write the docstring under {meth.__name__!r} and it will be "
+                        f"be passed to the {param_name!r} constructor automatically."
+                    )
+                if "name" in kwargs:
+                    raise RuntimeError(
+                        f"Failed to add parameter {param_name!r}.\n"
+                        f"`name` kwarg was provided to {meth.__name__!r} method. "
+                        f"The parameter name is obtained from {meth.__name__!r} name. "
+                        f"Do not specify it manually."
+                    )
+                self.add_parameter(
+                    name=param_name,
+                    docstring=meth.__doc__,
+                    **kwargs,
+                )
+
+
+class InstanceAttr:
+    """
+    An auxiliary class to be used together with the
+    :func:`@add_parameter <qcodes.instrument.base.add_parameter>` decorator to allow
+    adding parameters that require information available only during the
+    :code:`__init__()` of an Instrument subclass.
+    """
+    def __init__(self, attr_name: str) -> None:
+        """Instantiates the class and saves the passed in :code:`attr_name`."""
+        self.attr_name: str = attr_name
+
+    def __repr__(self) -> str:
+        """Returns a formatted string the with class name and :code:`self.attr_name`."""
+        return f"{self.__class__.__name__}({self.attr_name!r})"
+
+
+_DECORATED_METHOD_PREFIX = "_parameter_"
+"""
+A constant defining the prefix of the methods on which the
+:func:`@add_parameter <qcodes.instrument.base.add_parameter>` decorator can be used.
+The intention is to keep the name of these methods fairly unique and private to avoid
+any foreseeable clash.
+"""
+
+_ADD_PARAMETER_ATTR_NAME = "_add_parameter"
+"""
+A constant defining the name of the attribute set by the
+:func:`@add_parameter <qcodes.instrument.base.add_parameter>` decorator to flag a method
+that will be converted to parameter.
+"""
+
+_ParamSpec = TypeVar("_ParamSpec")
+"""
+A custom type to annotate the type hints of :obj:`qcodes.instrument.base.add_parameter`.
+"""
+
+def add_parameter(method: Callable[[_ParamSpec], Any]) -> Callable[[_ParamSpec], Any]:
+    r"""
+    A decorator function for adding parameters to instruments in the new style.
```

Comment:
let's avoid old/new style and be more specific
```suggestion
    A decorator function for adding parameters to instruments via method definitions.
```